zerver/lib/actions.py

```diff
@@ -4860,6 +4865,18 @@ def user_info(um: UserMessage) -> Dict[str, Any]:
 
             users_to_be_notified += list(map(subscriber_info, subscriber_ids))
 
+    if topic_name is not None:
+        users_muted_topic_data = get_users_muted_topic_data(orig_topic_name, stream_id)
+        for user_data in users_muted_topic_data:
+            user_profile = user_data["user_profile"]
+            remove_topic_mute(user_profile, stream_id, orig_topic_name)
+            add_topic_mute(user_profile, stream_id, user_data["recipient_id"],
+                           topic_name, user_data["date_muted"])
+            for user in users_to_be_notified:
+                if user_profile.id == user["id"]:
+                    user['muted_topics'] = get_topic_mutes(user_profile)
+                    break
+
```

Comment:
I feel it would be better to use the `do_mute_topic` and `do_unmute_topic` methods because-
- they will handle all the event sending work
- sending muted topic data in `update_messages` event seems a bit out of place
- you could avoid making any frontend changes at all, since we already handle `muted_topics` events (which these functions will send) 


Also, I feel this code-block can be optimized and made more readable --
```python
if topic_name is not None:
        for user in users_to_be_notified:
            if topic_is_muted(user, stream_id, orig_topic_name):
                # do the muting and unmuting
```
(You could avoid the backend `get_users_muted_topic_data` helper!)

Also, I don't think this is the correct place for this code block. If I'm not wrong, placing this code here would mute the new topic even if it -
1. Was moved to another stream? (I'm not sure if we want [that to happen](https://github.com/zulip/zulip/issues/15210#issuecomment-639156475))
2. Only that single message was edited, and not the whole topic (see `propagate_mode`)

I think this code belongs under some `if` statements.

Reply:
> I feel it would be better to use the `do_mute_topic` and `do_unmute_topic`

Calling `do_mute_topic` and `do_unmute_topic` for each user would lead to calling `send_event` function for each of the users which can rather be avoided if `muted_topic` data is sent along with `users_to_be_notified` (in final `send_event` function call which will be ultimately called to edit topic/messages)

> - sending muted topic data in `update_messages` event seems a bit out of place

`muted topic data` of each user will be ultimately handled by [this part of the code](https://github.com/zulip/zulip/blob/08d716c74175a8b63bf8eb1080381c77f3f853c6/zerver/tornado/event_queue.py#L960-L965).

> 3. you could avoid making any frontend changes at all, since we already handle `muted_topics` events (which these functions will send)

In that case, `muted_topic` event will be called twice, ultimately leading to repeated calls for some functions, at frontend, 2-3 times when `update_message` event will be sent.

>  If I'm not wrong, placing this code here would mute the new topic even if it -
> 1. Was moved to another stream? (I'm not sure if we want that to happen)

If a topic is moved to another stream then `get_users_muted_topic_data(orig_topic_name, stream_id)` function would take different `stream_id` as an argument, so that case will be easily avoided.

> Only that single message was edited, and not the whole topic (see `propagate_mode`)

I have tested this locally and it was working correctly.

Reply:
> muted topic data of each user will be ultimately handled by this part of the code.

Oh! I missed that, my bad. In any case, you could refactor the code as above, even if you don't use the `do_*` functions 🙂